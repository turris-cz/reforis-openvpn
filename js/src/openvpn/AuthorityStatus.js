/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React, { useEffect } from "react";
import PropTypes from "prop-types";

import { useAPIGet, Spinner } from "foris";

import API_URLs from "API";
import AuthorityMissing from "./AuthorityMissing";
import AuthorityReady from "./AuthorityReady";
import AuthorityBeingGenerated from "./AuthorityBeingGenerated";

AuthorityStatus.propTypes = {
    ws: PropTypes.object.isRequired,
};

export default function AuthorityStatus({ ws }) {
    const [authority, getAuthority] = useAPIGet(API_URLs.authority);
    useEffect(() => {
        getAuthority();
    }, [getAuthority]);

    if (authority.isLoading || !authority.data) {
        return <Spinner className="my-3 text-center" />;
    }

    if (authority.data.status === "missing") {
        return <AuthorityMissing onSuccess={getAuthority} />;
    }
    if (authority.data.status === "ready") {
        return <AuthorityReady onSuccess={getAuthority} />;
    }
    if (authority.data.status === "generating") {
        return <AuthorityBeingGenerated ws={ws} onSuccess={getAuthority} />;
    }
}
