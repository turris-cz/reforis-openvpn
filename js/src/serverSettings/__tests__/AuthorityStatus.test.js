/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";
import { render } from "foris/testUtils/customTestRender";

import { WebSockets } from "foris";

import AuthorityStatus from "../AuthorityStatus";

describe("<AuthorityStatus />", () => {
    const webSockets = new WebSockets();

    function renderAuthorityStatus(status) {
        const { container } = render(
            <AuthorityStatus
                ws={webSockets}
                status={status}
                serverEnabled={false}
                onReload={jest.fn()}
            >
                <div>Some child</div>
            </AuthorityStatus>
        );
        return container;
    }

    it("should handle missing CA", () => {
        const container = renderAuthorityStatus("missing");
        expect(container).toMatchSnapshot();
    });

    it("should handle ready CA", () => {
        const container = renderAuthorityStatus("ready");
        expect(container).toMatchSnapshot();
    });

    it("should handle generating CA", () => {
        const container = renderAuthorityStatus("generating");
        expect(container).toMatchSnapshot();
    });

    it("should render children if status is ready", () => {
        const { getByTestId } = render(
            <AuthorityStatus
                ws={webSockets}
                status={"ready"}
                serverEnabled={false}
                onReload={jest.fn()}
            >
                <div data-testid="some-child">
                    Status is "ready" so I should be visible!
                </div>
            </AuthorityStatus>
        );
        expect(getByTestId("some-child")).toBeDefined();
    });
});
