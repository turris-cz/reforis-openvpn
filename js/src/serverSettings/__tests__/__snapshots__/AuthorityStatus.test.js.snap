// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AuthorityStatus /> should handle generating CA 1`] = `
<div>
  <div
    class="card p-4"
  >
    <h2>
      Certificate Authority
    </h2>
    <p>
      Your certification authority is now being generated. It usually takes a few minutes. Settings will appear here automatically once the authority is ready.
    </p>
    <div
      class="text-center text-muted"
    >
      <div
        class="spinner-border spinner-border-sm me-1"
        role="status"
      >
        <span
          class="sr-only"
        />
      </div>
      Generating certification authorityâ€¦
    </div>
  </div>
</div>
`;

exports[`<AuthorityStatus /> should handle missing CA 1`] = `
<div>
  <div
    class="card p-4"
  >
    <h2>
      Certificate Authority
    </h2>
    <p>
      Currently there is no OpenVPN certificate authority (CA). A CA is required to generate client certificates to authenticate to the OpenVPN server. To proceed you need to generate it first.
    </p>
    <p
      class="text-center text-muted"
    >
      No certification authority.
    </p>
    <div
      class="text-end"
    >
      <button
        class="btn btn-primary col-12 col-md-3 col-lg-2 d-inline-flex justify-content-center align-items-center"
        type="button"
      >
        <span>
          Generate CA
        </span>
      </button>
    </div>
  </div>
</div>
`;

exports[`<AuthorityStatus /> should handle ready CA 1`] = `
<div>
  <div>
    Some child
  </div>
  <div
    class="card p-4 col-sm-12 col-lg-12 p-0 mb-4"
  >
    <h2>
      Certificate Authority
    </h2>
    <p>
      Your certificate authority (CA) is set up properly. Please note that if you delete it, all clients will have their access revoked.
    </p>
    <div
      class="text-end"
    >
      <button
        class="btn btn-danger col-12 col-md-3 col-lg-2 d-inline-flex justify-content-center align-items-center"
        type="button"
      >
        <span>
          Delete CA
        </span>
      </button>
    </div>
  </div>
</div>
`;
